File: scala-project/src/main/scala/com/waitingforcode/Enricher.scala
package com.waitingforcode

import org.apache.spark.sql.{DataFrame, Dataset, functions => F}

object Enricher {

  def enrichSessionsWithDevices(sessions: Dataset[Session], devices: DataFrame): DataFrame = {
    val enrichedVisits = sessions.join(
      devices,
      sessions("device_type") === F.col("type") &&
      sessions("device_version") === F.col("version"),
      "left_outer"
    ).drop("type", "version", "device_full_name")
      .withColumnRenamed("full_name", "device_full_name")

    enrichedVisits
  }

}

File: python-project/visits_generator/enricher.py
from pyspark.sql import DataFrame, functions


def enrich_sessions_with_devices(sessions: DataFrame, devices: DataFrame) -> DataFrame:
    enriched_visits = ((sessions.join(
        devices,
        [sessions.device_type == devices.type,
         sessions.device_version == devices.version],
        'left_outer')).drop('type', 'version', 'device_full_name')
                       .withColumnRenamed('full_name', 'device_full_name'))

    return enriched_visits
